[30585] Extract copying projects in to service and add boards / manual sorting

oliverguenther opened the pull request:
- Refactors `Project#copy` into a `CopyService` that extends from a base service that should allow copying models with associations more easy. The base service rents out to multiple `Dependency` services that copy one association and do error handling for that one assocation.
- State can be passed between dependent services (such as the work_packages_map for work packages to queries) through the use of a `Shared::ServiceState` object, which is just an openstruct that gets passed to the service and back to the result.
- The memorized state for `work_packages_map` has been reduced to an integer instead of whole work packages, which was not really needed and should reduce memory usage for larger projects
- Adds the copying of manually sorted orders to the queries dependency service
- Adds the copying of boards to the project copy service
- Converts legacy project spec examples and model project_copy specs that were related to copying into the newly created `copy_service_integration_spec`

Note:  Includes what was previously https://github.com/opf/openproject/pull/8491 . I closed the PR in favor of this one that refactors the `state:` argument to the services to something more robust

https://community.openproject.com/wp/30585

oliverguenther committed:
Refactor copy project into service

oliverguenther committed:
Retain order in copied work packages if query is manually sorted

oliverguenther committed:
Move legacy project copy spec to service spec

oliverguenther committed:
Ensure reform error merge syntax is compatible with ActiveModel

oliverguenther committed:
Retain the public attribute AFTER the SetAttributeservice

The public attribute gets reset in the SetAttributeService

oliverguenther committed:
Add spec for manually sort order

oliverguenther committed:
Refactor copy services to use shared state

oliverguenther committed:
Avoid definition of acts_as_list

This is going to reorder items as they are being inserted, and we don't even use
acts_as_list for ordered work packages

oliverguenther committed:
Add test for base callable service

oliverguenther committed:
Correctly create board queries so they are manual sorted

If they do not get the sort criteria applied, the test will not assume they are manually sorted

oliverguenther committed:
Add spec for copying of boards

oliverguenther committed:
Correctly base all services on base_contracted with a model

BaseContracted service uses a `model` reference that it doesn't define
itself, neither does some of the dependent services.

To avoid this, make sure the BaseContracted gets an optional model
attribute that can be set.

oliverguenther committed:
Fix params[:only] check for wiki dependent service

oliverguenther committed:
Refactor archive service to use model from BaseContract

oliverguenther committed:
Add common \`should_copy?\` method for the project copy service

oliverguenther committed:
Disable notifications for user in the bsae copy service

ulferts commented on app/services/projects/copy/members_dependent_service.rb:
```diff
@@ -0,0 +1,55 @@
+        new_member.role_ids = role_ids
+        new_member.project = target
+        target.memberships << new_member
+        new_member.save
```
This line does nothing given the line above. The assignment to the already saved target's memberships association will already have saved the new member. 

I am aware that this is probably copied code. But it might be a good chance to improve it.

I feel that the culprit here is the magical association assignment which depending on the state of target will either only assign or create.

It might be my preference but I think something along

```
role_ids = member.member_roles.reject(&:inherited?).map(&:role_id)

next if role_ids.empty?

attributes = member.attributes.dup.except('id', 'project_id', 'created_on'))
            .merge(role_ids: role_ids)
                    
target.memberships.create attributes
```

is easier to read. The pattern above would lead to other dependent_services e.g. for version https://github.com/opf/openproject/pull/8498/files#diff-bcb0f264837b6701438da5957a9cb803R37-R39 being handled in one statement

```
target.versions.create  version.attributes.dup.except('id', 'project_id', 'created_on', 'updated_at')
```

oliverguenther replied:
Looks good to me. I did not touch any of the copying code but will modify this

ulferts commented on app/services/projects/copy/work_packages_dependent_service.rb:
```diff
@@ -0,0 +1,157 @@
+
+    def work_package_responsible(source_work_package)
+      source_work_package.responsible && target.possible_responsibles.detect { |u| u.id == source_work_package.responsible_id }
+    end
```
Those lookup will run repeatedly, for every work package to copy. Transforming it into a hash that is kept between lookups would be way faster.

oliverguenther replied:
For versions, we can even use a lookup map as they got created before. Will do that

ulferts commented on app/services/queries/copy_service.rb:
```diff
@@ -0,0 +1,56 @@
+    end
+
+    def initialize_copy(source, params)
+      new_query = ::Query.new name: '_'
```
Why provide a name if it is overwritten in the next line?

oliverguenther replied:
Again that was simply copied from the `copy_queries` method. I adjusted that accordingly

ulferts reviewed (changes requested):
I like the structure you applied to the copying process. It should it make it easier to improve individual shortcomings of the process.

Copying sorted work packages in a query works fine except for work packages having a parent that is not copied along as it is in a different project:

![image](https://user-images.githubusercontent.com/617519/87529254-0386a480-c68f-11ea-91be-09206e46cf83.png)

oliverguenther committed:
Make dependent services use project.(association).create when possible

oliverguenther committed:
Improve lookup of versions and work package user assignments

oliverguenther committed:
Return the unmapped work package ID in case of wp not copied

When a foreign work package is linked, it will not get copied during the
project copy service. In this case, just use the original to retain it.

oliverguenther commented:
@ulferts I believe to have addressed all remarks. Turns out that the project wasn't always saved in the `initialize_copy` step so far so I added an explicit save + return in case of errors here: https://github.com/opf/openproject/pull/8498/commits/2523040abf67601f0a8d3e39016d0be8bf9bb0fb#diff-5482eade3aefd326f78ea675a1a14551R57
